### PR TITLE
Allow nesting CONTAIN predicates with and/or

### LINKED ISSUE #2188: "within" step combined with "or" step throws errors
- Version: 0.5.2
- Storage Backend: Scylla
- Mixed Index Backend: Elasticsearch
- Mailing list Thread URL: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/janusgraph-users/-tSt8U1kvxA/xz95la6DAgAJ
- Steps to Reproduce: 
We have a schema with a property of type "Long", when we try to execute simple Gremlin query in the console

g.V().or(__.has("prop",P.within(2L)),__.has("prop",P.within(1L)))

We get the error: 

Value [[2]] is not an instance of the expected data type for property key [prop] and cannot be converted. Expected: class java.lang.Long, found: class java.util.Arrays$ArrayList

While g.V().has("prop",P.within()) can be executed just fine
### Stack Trace 

```
2020-08-17 11:21:41.310 -  [gremlin-server-exec-6] WARN  o.a.t.g.s.op.AbstractEvalOpProcessor - Exception processing a script on request [RequestMessage{, requestId=153f5014-2d6a-4aa9-934d-43d506cdb0c6, op='eval', processor='', args={gremlin=g.V().or(__.has("prop",P.within(2L)),__.has("prop",P.within(1L))), bindings={}, batchSize=64}}].
org.janusgraph.core.SchemaViolationException: Value [[2]] is not an instance of the expected data type for property key [prop] and cannot be converted. Expected: class java.lang.Long, found: class java.util.Arrays$ArrayList
	at org.janusgraph.graphdb.transaction.StandardJanusGraphTx.verifyAttribute(StandardJanusGraphTx.java:578)
	at org.janusgraph.graphdb.query.QueryUtil.addConstraint(QueryUtil.java:267)
	at org.janusgraph.graphdb.query.QueryUtil.addConstraint(QueryUtil.java:257)
	at org.janusgraph.graphdb.query.QueryUtil.constraints2QNF(QueryUtil.java:209)
	at org.janusgraph.graphdb.query.graph.GraphCentricQueryBuilder.constructQueryWithoutProfile(GraphCentricQueryBuilder.java:238)
	at org.janusgraph.graphdb.query.graph.GraphCentricQueryBuilder.constructQuery(GraphCentricQueryBuilder.java:225)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.buildGraphCentricQuery(JanusGraphStep.java:148)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.lambda$null$0(JanusGraphStep.java:89)
	at java.base/java.util.LinkedHashMap$LinkedEntrySet.forEach(LinkedHashMap.java:671)
	at org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphStep.lambda$new$2(JanusGraphStep.java:89)
	at org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep.processNextStart(GraphStep.java:157)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
	at org.apache.tinkerpop.gremlin.process.traversal.util.DefaultTraversal.hasNext(DefaultTraversal.java:197)
	at org.apache.tinkerpop.gremlin.server.op.AbstractOpProcessor.handleIterator(AbstractOpProcessor.java:93)
	at org.apache.tinkerpop.gremlin.server.op.AbstractEvalOpProcessor.lambda$evalOpInternal$5(AbstractEvalOpProcessor.java:264)
	at org.apache.tinkerpop.gremlin.groovy.engine.GremlinExecutor.lambda$eval$0(GremlinExecutor.java:278)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by rngcntr**
Fixes #2188

The needed changes are already described in #2188. To solve this issue, it was also necessary to alter the definition of two `PredicateCondition`s being equal to each other. Nesting `P.without()` predicates on a property value within an `and()` step automatically adds a `PredicateCondition` that this property value must not equal `null`. Therefore, `PredicateCondition` should handle `null` values on equality checks.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java [hunk 1]
```diff
@@ -241,7 +262,29 @@ public static boolean isEmpty(Condition<?> condition) {
...
+                    }
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
+                } else {
+                    Preconditions.checkArgument(janusGraphPredicate == Contain.IN);
+                    if (childValues.isEmpty()) {
+                        return null; //Cannot be satisfied
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:281]: This would fail the test
```java
assertEquals(3, g.V().or(__.has("id", P.within(2, 3, 4)), __.has("id", P.within())).count().next());
```

**Review by li-boxuan — COMMENTED** [review]
Nice and quick work! I think you may want to add more tests for edge cases (like the one I mentioned below), as this part is tricky and error-prone.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java (hunk 1 above)
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java]: Thanks for the review. Basically needed to change two lines within the definition of `OR` to catch that bug. I updated the branch by pushing a separate commit containing the fix so it is easier to review.

Also added 14 new test cases which cover all of the empty argument cases as well as some previously untested combinations of `with` and `without` cases. If you can find different combinations worth testing, please let me know.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java [hunk 2]
```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
...
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    // Don't need to handle the case where childValues is empty, because it defaults to
+                    // an or(and()) is added, which is a tautology
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:278]: I wonder if there is any performance difference between:

`OR(NOT_EQUAL(x) AND NOT_EQUAL(y))`
v.s.
`OR(NOT_EQUAL(x)) OR(NOT_EQUAL(y))`?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java [hunk 3]
```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
...
-            if (janusGraphPredicate instanceof AndJanusPredicate) {
+            if (janusGraphPredicate instanceof Contain) {
+                //Rewrite contains conditions
+                final Collection childValues = (Collection) values.get(i);
+                if (janusGraphPredicate == Contain.NOT_IN) {
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    // Don't need to handle the case where childValues is empty, because it defaults to
+                    // an or(and()) is added, which is a tautology
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:273]: I understand they are semantically the same, but will there be a performance difference (even if very little)?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java [hunk 4]
```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
...
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
+                } else {
+                    Preconditions.checkArgument(janusGraphPredicate == Contain.IN);
+                    if (childValues.isEmpty()) {
+                        continue; // Handle unsatisfiable condition within or like it does not exist
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:282]: nitpick: it took me a while to understand the comment. I guess you mean "Handle unsatisfiable condition `or(within())` like it does not exist"

**Review by li-boxuan — COMMENTED** [review]
The logic and tests look good to me, just left a few comments.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java (hunk 3 above)
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:273]: There may be a little difference in performance, but the empty `and` actually serves a functional purpose. Otherwise the surrounding `or` can be empty which results in `or()` being evaluated as `false` instead of `or(and())` which is `true`.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java (hunk 2 above)
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:278]: I don't quite get your point here. It looks like your second case is missing a logical operator between the two `OR`s. Anyway, I can't spot what kind of possible change in the code you refer to, in order to create one or the other of your examples.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java (hunk 4 above)
Comment (rngcntr) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:282]: It was actually meant to say "Handle any unsatisfiable condition within an OR statement like it does not exist."

**Comment by rngcntr** [comment]
Notice: One could argue that this is a bugfix which should also be applied to v0.5.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java (hunk 2 above)
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java:278]: Oh never mind, I got it wrong

**Review by li-boxuan — APPROVED** [review]
LGTM 👍  Can you squash the commits before we merge it?

**Comment by rngcntr** [comment]
@li-boxuan Sure.
Does anybody know if I can use Travis to ensure everything works on v0.5 as well? Or do I have to run the complete test set locally?

**Comment by li-boxuan** [comment]
> Does anybody know if I can use Travis to ensure everything works on v0.5 as well

I don't know if it's possible to run it under janusgraph namespace, but I think you can merge it to v0.5 and push to your own fork first. Travis should be able to run tests on that branch automatically. If not, you can anyways trigger Travis manually.

**Comment by rngcntr** [comment]
Thanks @li-boxuan!

I restarted the build on Travis for this PR yesterday because there was a failing test which passed before squashing. Now that one passes but another one fails. This is exhausting!

**Comment by li-boxuan** [comment]
Changing milestone because it was cherry-picked to v0.5 branch: https://github.com/JanusGraph/janusgraph/commit/7ad1d37a04512621320373a2f36fcf02bfed8297
